snuba/utils/streams/kafka.py

```diff
@@ -46,34 +77,76 @@ class KafkaConsumer(Consumer[TopicPartition, int, bytes]):
 
     def __init__(self, configuration: Mapping[str, Any]) -> None:
         self.__consumer = ConfluentConsumer(configuration)
-
-    def __wrap_assignment_callback(
-        self, callback: Callable[[Sequence[TopicPartition]], None]
-    ) -> Callable[[ConfluentConsumer, Sequence[ConfluentTopicPartition]], None]:
-        def wrapper(
-            consumer: ConfluentConsumer, partitions: Sequence[ConfluentTopicPartition]
-        ) -> None:
-            callback(
-                [TopicPartition(value.topic, value.partition) for value in partitions]
-            )
-
-        return wrapper
+        self.__auto_offset_reset = get_enum_configuration_value(
+            configuration,
+            "auto.offset.reset",
+            {
+                "earliest": set(["smallest", "beginning"]),
+                "latest": set(["largest", "end"]),
+                "error": set(),
+            },
+            "largest",
+        )
+        self.__offsets: MutableMapping[TopicPartition, int] = {}
+        self.__state: ConsumerState = ConsumerState.CONSUMING
 
     def subscribe(
         self,
         topics: Sequence[str],
         on_assign: Optional[Callable[[Sequence[TopicPartition]], None]] = None,
         on_revoke: Optional[Callable[[Sequence[TopicPartition]], None]] = None,
     ) -> None:
-        kwargs = {}
+        def assignment_callback(
+            consumer: ConfluentConsumer, partitions: Sequence[ConfluentTopicPartition]
+        ) -> None:
+            self.__state = ConsumerState.ASSIGNING
+
+            # TODO: This might be able to use the ``positions`` API to avoid
+            # rolling back offsets after an uncommitted rebalance?
+            offsets: MutableMapping[TopicPartition, int] = {}
+
+            for i in self.__consumer.committed(partitions):
+                stream = TopicPartition(i.topic, i.partition)
+                if i.offset >= 0:
+                    offsets[stream] = i.offset
+                elif i.offset == OFFSET_INVALID:
+                    # TODO: What arguments should be provided to ``get_watermark_offsets``?
```

Comment:
This has the option of being synchronous or asynchronous: https://docs.confluent.io/current/clients/confluent-kafka-python/#confluent_kafka.Consumer.get_watermark_offsets

The latest value should be pretty up to date, while the earliest value sounds like it might be somewhat stale. The only concern with reading a cached earliest value is that might potentially be stale enough that it's been since evicted and a subsequent `poll` call will raise an offset out of range error. We use this uncached in the `SynchronizedConsumer` and to my knowledge haven't seen any issues with it but I don't think we have any specific metrics on.

My main concern here would be rebalancing is a relatively time-sensitive operation and a failed rebalance can throw the cluster into a weird state until all of the consumers converge on assignments again so making this as smooth and reliable as possible is probably in our best interest.